### PR TITLE
Add semantic version operators for feature flag evaluation

### PR DESCRIPTION
**Pull request by greghuels**
## What Does This Do

Ports SemVer condition operators (`SEMVER_EQ`, `SEMVER_NEQ`, `SEMVER_LT`, `SEMVER_LTE`, `SEMVER_GT`, `SEMVER_GTE`) from [dd-trace-go#5128](https://github.com/DataDog/dd-trace-go/pull/5128) to the Java SDK.

- **`ParsedSemver`** — Rust-compatible SemVer parser and comparator (unsigned 64-bit core components, arbitrary-length numeric prerelease identifiers, build metadata validated but ignored for precedence).
- **`ConditionOperator`** — 6 new enum values.
- **`ConditionConfiguration`** — transient `semverComparand` field cached during config validation.
- **`ServerConfiguration`** — transient `invalidFlags` map so the evaluator returns `PARSE_ERROR` for flags with invalid semver comparands.
- **`DDEvaluator`** — `evaluateSemverCondition` method and switch cases for all 6 operators; invalid-flag check returning `PARSE_ERROR`.
- **`UniversalFlagConfigParser`** — validates and caches semver comparands during parsing; invalid comparands cause the flag to be dropped and tracked.
- **`ffe-system-test-data` submodule** — updated to version that included the semver tests (`aaa97e4`) which includes [ffe-system-test-data#20](https://github.com/DataDog/ffe-system-test-data/pull/20) and [ffe-system-test-data#22](https://github.com/DataDog/ffe-system-test-data/pull/22).

## Motivation

Cross-SDK feature flag parity. The Go SDK added SemVer operators in dd-trace-go#5128; this ports the same functionality to Java so FFE targeting rules using `SEMVER_*` operators evaluate identically across SDKs.

## Additional Notes

- Unit tests: `ParsedSemverTest` (30 tests ported from Go `semver_test.go`), `DDEvaluatorTest` semver condition tests (28 parameterized + 2 edge cases).
- Canonical fixture tests: 22 cases from the two system test PRs run automatically via the existing `testEvaluateCanonicalFixture` parameterized test and the `OpenFeatureProviderSmokeTest`.
- All 551 unit tests and 250 smoke tests pass; Spotless and SpotBugs clean.

## Contributor Checklist

- [x] Format the title according to [the contribution guidelines](https://github.com/DataDog/dd-trace-java/blob/master/CONTRIBUTING.md#title-format)
- [x] Assign the `type:` and (`comp:` or `inst:`) labels in addition to [any other useful labels](https://github.com/DataDog/dd-trace-java/blob/master/CONTRIBUTING.md#labels)
- [x] Avoid using `close`, `fix`, or [any linking keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) when referencing an issue
- [x] Update the [CODEOWNERS](https://github.com/DataDog/dd-trace-java/blob/master/.github/CODEOWNERS) file on source file addition, migration, or deletion
- [x] Update [public documentation](https://docs.datadoghq.com/tracing/trace_collection/library_config/java/) with any new configuration flags or behaviors
- [ ] Once approved, [use merge queue](https://github.com/DataDog/dd-trace-java/blob/master/CONTRIBUTING.md#merge-queue) to merge the PR

Jira ticket: [FFL-2920]

[FFL-2920]: https://datadoghq.atlassian.net/browse/FFL-2920?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: products/feature-flagging/feature-flagging-api/src/main/java/datadog/trace/api/openfeature/DDEvaluator.java [hunk 1]
```diff
@@ -318,16 +358,49 @@ private static boolean compareNumber(
...
+    return comparator.compare(ParsedSemver.compare(parsedAttribute, comparand));
+  }
+
   private static boolean matchesShard(final Shard shard, final String targetingKey) {
-    final int assignedShard = getShard(shard.salt, targetingKey, shard.totalShards);
+    if (shard.totalShards <= 0 || shard.totalShards > 0xFFFFFFFFL) {
+      throw new InvalidFlagConfigException(
+          "Invalid shard totalShards (must be in unsigned 32-bit range): " + shard.totalShards);
+    }
+    final int assignedShard = getShard(shard.salt, targetingKey, (int) shard.totalShards);
```
[the remark this answers is not in the document]
Comment (greghuels) [reply products/feature-flagging/feature-flagging-api/src/main/java/datadog/trace/api/openfeature/DDEvaluator.java]: Good catch — fixed in b2f1dabea9. Changed `getShard` to accept `long totalShards` and removed the `(int)` cast at the call site. Values in the upper half of the unsigned 32-bit range (e.g. `0xFFFFFFFFL`) were being cast to negative ints, making `intFromHash % totalShards` always return 0 and causing every targeting key to match shard range 0. The modulo now operates on the original `long` value.

File: products/feature-flagging/feature-flagging-api/src/main/java/datadog/trace/api/openfeature/DDEvaluator.java [hunk 2]
```diff
@@ -111,6 +112,10 @@ public <T> ProviderEvaluation<T> evaluate(
 
       final Flag flag = config.flags.get(key);
       if (flag == null) {
+        if (config.invalidFlags != null && config.invalidFlags.containsKey(key)) {
+          return error(
+              defaultValue, ErrorCode.PARSE_ERROR, "invalid configuration for flag " + key);
```
[the remark this answers is not in the document]
Comment (greghuels) [reply products/feature-flagging/feature-flagging-api/src/main/java/datadog/trace/api/openfeature/DDEvaluator.java]: This is intentional. The updated `ffe-system-test-data` fixture includes flags with various malformations (invalid operators, non-list allocations, etc.), not just invalid semver comparands. If a flag was present in the config but dropped by the parser due to any parse error, returning `FLAG_NOT_FOUND` would be misleading — the flag exists but is malformed. `PARSE_ERROR` correctly communicates that the flag's configuration is invalid, which is what the evaluation test cases in the fixture expect.

File: products/feature-flagging/feature-flagging-api/src/main/java/datadog/trace/api/openfeature/DDEvaluator.java [hunk 3]
```diff
@@ -246,10 +253,13 @@ private static boolean evaluateRules(final List<Rule> rules, final EvaluationCon
   private static boolean evaluateCondition(
       final ConditionConfiguration condition, final EvaluationContext context) {
     if (condition.operator == ConditionOperator.IS_NULL) {
+      if (!(condition.value instanceof Boolean)) {
+        throw new InvalidFlagConfigException(
+            "IS_NULL condition requires a boolean value, got: " + condition.value);
+      }
```
[the remark this answers is not in the document]
Comment (greghuels) [reply products/feature-flagging/feature-flagging-api/src/main/java/datadog/trace/api/openfeature/DDEvaluator.java]: This is intentional per the updated fixture suite. The `IS_NULL` operator's `value` field specifies whether to check for null (`true`) or non-null (`false`). A non-boolean `value` is a malformed configuration, not a missing field — the UFC schema specifies `value` as a boolean for `IS_NULL` conditions. Throwing `InvalidFlagConfigException` → `PARSE_ERROR` is the correct behavior for schema-invalid configs. The evaluation test cases include such malformed conditions and expect `PARSE_ERROR`.

File: products/feature-flagging/feature-flagging-api/src/main/java/datadog/trace/api/openfeature/DDEvaluator.java [hunk 4]
```diff
@@ -176,10 +183,10 @@ public <T> ProviderEvaluation<T> evaluate(
           .value(defaultValue)
           .reason(Reason.DEFAULT.name())
           .build();
-    } catch (final PatternSyntaxException e) {
+    } catch (final PatternSyntaxException | InvalidFlagConfigException e) {
       return error(defaultValue, ErrorCode.PARSE_ERROR, e);
     } catch (final NumberFormatException e) {
-      return error(defaultValue, ErrorCode.TYPE_MISMATCH, e);
+      return error(defaultValue, ErrorCode.PARSE_ERROR, e);
```
[the remark this answers is not in the document]
Comment (greghuels) [reply products/feature-flagging/feature-flagging-api/src/main/java/datadog/trace/api/openfeature/DDEvaluator.java]: This is intentional. When a numeric condition (e.g. `GT 18`) receives a non-numeric context value like `"unknown"`, the `NumberFormatException` comes from parsing the **context attribute**, not from comparing it. The condition's configuration (operator + comparand) is valid — the issue is that the caller supplied an unparseable value. However, the updated fixture suite expects `PARSE_ERROR` for these cases because the flag's evaluation cannot proceed with the given context type, and `TYPE_MISMATCH` in OpenFeature is specifically for when the resolved value type doesn't match the expected type (not for unparseable context attributes). The test cases in the fixture validate this behavior.

File: products/feature-flagging/feature-flagging-api/src/main/java/datadog/trace/api/openfeature/DDEvaluator.java [hunk 5]
```diff
@@ -420,7 +493,9 @@ private static <T> ProviderEvaluation<T> resolveVariant(
             .reason(
                 !isEmpty(allocation.rules)
                     ? Reason.TARGETING_MATCH.name()
-                    : !isEmpty(split.shards) ? Reason.SPLIT.name() : Reason.STATIC.name())
+                    : !isEmpty(split.shards)
+                        ? Reason.SPLIT.name()
+                        : isDateFiltered(allocation) ? Reason.DEFAULT.name() : Reason.STATIC.name())
```
[the remark this answers is not in the document]
Comment (greghuels) [reply products/feature-flagging/feature-flagging-api/src/main/java/datadog/trace/api/openfeature/DDEvaluator.java]: This is intentional. When an allocation is date-filtered (has `startAt`/`endAt` but no rules and no shards), the evaluation falls through to the flag's default variation — but only because the date gate allowed it through, not because a specific rule or shard matched. Using `DEFAULT` here distinguishes "served the default variation due to no active targeting rule" from `STATIC` which implies a deterministic, always-on assignment. The fixture's evaluation test cases expect `DEFAULT` for date-gated allocations with no matching rules or splits.

File: products/feature-flagging/feature-flagging-bootstrap/src/main/java/datadog/trace/api/featureflag/ufc/v1/ParsedSemver.java [hunk 6]
```diff
@@ -0,0 +1,316 @@
+package datadog.trace.api.featureflag.ufc.v1;
+
+/**
+ * ParsedSemver is the language-neutral representation of the Rust/Eppo SemVer subset used by FFE.
```
Comment (leoromanovsky) [inline products/feature-flagging/feature-flagging-bootstrap/src/main/java/datadog/trace/api/featureflag/ufc/v1/ParsedSemver.java]: 👀

File: products/feature-flagging/feature-flagging-lib/src/main/java/com/datadog/featureflag/UniversalFlagConfigParser.java [hunk 7]
```diff
@@ -58,14 +70,92 @@ public ServerConfiguration deserialize(final byte[] content) throws IOException
...
+   * Validates and caches SemVer comparands for all SEMVER_* conditions in a flag. Throws {@link
+   * InvalidSemverComparandException} if any condition has an invalid or non-string comparand value.
+   */
+  private static void validateAndCacheSemverComparands(final String flagKey, final Flag flag) {
+    if (flag.allocations == null) {
+      return;
+    }
+    for (int allocIdx = 0; allocIdx < flag.allocations.size(); allocIdx++) {
+      final Allocation allocation = flag.allocations.get(allocIdx);
+      if (allocation.rules == null) {
```
[the remark this answers is not in the document]
Comment (leoromanovsky) [reply products/feature-flagging/feature-flagging-lib/src/main/java/com/datadog/featureflag/UniversalFlagConfigParser.java:121]: We'll address this in a follow-up PR; more ffe fixture updates coming.

File: products/feature-flagging/feature-flagging-api/src/main/java/datadog/trace/api/openfeature/DDEvaluator.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (leoromanovsky) [reply products/feature-flagging/feature-flagging-api/src/main/java/datadog/trace/api/openfeature/DDEvaluator.java]: Only pull in updates from the git submodule that include your semver fixtures; I'll address the flag invalidate ones in a follow up.

[the remark this answers is not in the document]
Comment (greghuels) [reply products/feature-flagging/feature-flagging-api/src/main/java/datadog/trace/api/openfeature/DDEvaluator.java]: updated

**Review by leoromanovsky — APPROVED** [review]
